### DISCUSSION TITLE
Question about the order of LSP requests and notifications (didChange and semanticTokens)

### DISCUSSION THREAD

**Discussion by He1pa** [microsoft/vscode-discussions#1219] [Extension Development QnA] [answered]
I am a developer of the lsp language extension. I noticed that under normal circumstances, when a user enters code in vscode, lsp will:
1 Send a notification of file changes (Notification { method: "textDocument/didChange"})
2 Request new semantic highlighting (Request { method: "textDocument/semanticTokens})

But I found that when inputting continuously and quickly (such as typing two `Enter` or other keys), sometimes semantic highlighting will be requested first, and then file changes will be notified.

I'm not sure if this is a problem with my lsp client implementation, or a special design of vscode, or a configurable option of vscode

here is my language extension:
client: https://github.com/kcl-lang/vscode-kcl
server: https://github.com/kcl-lang/kcl/tree/main/kclvm/tools/src/LSP

**Comment by tamuratak** [comment, marked as answer]
It sounds like a bug of `vscode-languageclient`. A similar issue can be found microsoft/vscode-languageserver-node/issues/1025. You had better report to https://github.com/microsoft/vscode-languageserver-node/issues.
